### PR TITLE
Client filters researcher

### PR DESCRIPTION
**Pull request by barbarah**
This pull request is the start of the shared filter logic for the Dataset Browser and the Researcher.

- I have created a Zustand store in the new 'list-store' package.
- I have refactored the client-filters component, so it uses the list store and removed dataset-specific logic (mainly the filters).
- I have moved one function of `lib/search-params` (getUrlWithSearchParams) to the store. The other functions are more project-specific and may need to be moved, but this takes more time.
- I have duplicated the UI components for the Researcher app.

Nest steps, *out of scope* for this pull request:

- I want to refactor the client-filters component and his children and move to the UI package.
    - Use the store in all components instead of passing down the state. Logic can also move to the store, leaving only UI in the client components.
    - Change to compound components with dot notation. With compound components, it is easier to have UI differences in the apps.

### CODE REVIEW

**Comment by barbarah** [comment]
Not working after the 'Next' upgrade to 13.3.1.  For some reason, the initialization in `client-list-store.tsx ` is not going correctly which results in falling end-to-end tests. After typing in the query field, the query will reset with an old value. This does not happen with `13.2.4`.

File: packages/list-store/src/useListStore.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  sortBy: SortBy;
+  // Setting newDataNeeded to true will trigger a page reload with new search params
+  newDataNeeded: boolean;
+  initialized: boolean;
+  selectedFilters: {[filterKey: string]: string[] | undefined};
+  setSelectedFilters: (key: string, value: string[]) => void;
+  setSortBy: (sortBy: SortBy) => void;
+  setQuery: (query: string) => void;
+  setPage: (direction: 1 | -1) => void;
+  composed: {
```
Comment (sdevalk) [inline packages/list-store/src/useListStore.ts:19]: I see where you want to go with this. But, to me, this is also a bit hard to follow. For example: what is the meaning of `composed` - why and when would I need this (if I were to use this store/hook)? Do you see a way to make this clear, or to simplify the implementation?

File: apps/researcher/src/app/[locale]/paginator.tsx
Comment (sdevalk) [inline apps/researcher/src/app/[locale]/paginator.tsx:1]: The component looks identical to the one of the Dataset Browser. Is there a way to re-use it?

File: apps/researcher/src/app/[locale]/selected-filters.tsx
Comment (sdevalk) [inline apps/researcher/src/app/[locale]/selected-filters.tsx:1]: The component looks identical to the one of the Dataset Browser. Is there a way to re-use it?

Comment (barbarah) [reply apps/researcher/src/app/[locale]/selected-filters.tsx:1]: Yes. For now, I placed this out of scope and only focused on subtracting logic. I put this in the pull request description, but it's properly unclear which components I'm talking about. I will create a ticket and update the description of this pull request.

File: apps/researcher/src/app/[locale]/paginator.tsx
Comment (barbarah) [reply apps/researcher/src/app/[locale]/paginator.tsx:1]: See response: https://github.com/colonial-heritage/colonial-collections/pull/38#discussion_r1179991561

File: packages/list-store/package.json [hunk 2]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "list-store",
```
[the remark this answers is not in the document]
Comment (barbarah) [reply packages/list-store/package.json:2]: Ah, I fought you did this for potentially sharing the packages and at this point, I was not planning to publish this package. 

The main reason I did not do this was because I didn't like the long names. I write all my terminal commands on the root level.  So if I need to install a new package, I would need to do `npm install newPackage -w @colonial-collections/list-store`. But this is properly not a good argument, so I will rename the package.

File: packages/list-store/src/useListStore.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+export interface Props {
+  totalCount: number;
+  offset: number;
+  limit: number;
+  query: string;
+  sortBy: SortBy;
+  // Setting newDataNeeded to true will trigger a page reload with new search params
+  newDataNeeded: boolean;
+  initialized: boolean;
```
[the remark this answers is not in the document]
Comment (barbarah) [reply packages/list-store/src/useListStore.ts:13]: I agree. I will prefix with `is`

File: packages/list-store/src/useSearchParamsUpdate.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
+'use client';
+
+import {useRouter} from 'next-intl/client';
+import {useEffect, useTransition} from 'react';
+import {useListStore} from './useListStore';
+
+export function useSearchParamsUpdate() {
+  const router = useRouter();
+  // Use the first param `isPending` of `useTransition` for a loading state.
```
[the remark this answers is not in the document]
Comment (barbarah) [reply packages/list-store/src/useSearchParamsUpdate.tsx:9]: I will remove this comment. The comment was there because the code used to be `const [, startTransition] = useTransition();` I wanted to explain the first empty param. Especially since we did need this param, we wanted to add a loading state.

File: packages/list-store/src/useListStore.ts (hunk 1 above)
Comment (barbarah) [reply packages/list-store/src/useListStore.ts:19]: I'm used to using a state manager with the possibility of derived values. (https://mobx-state-tree.js.org/concepts/views). I was trying to do the same, but I can see that this is unclear. I think it is better to move this entire logic to `useSearchParamsUpdate.tsx`. This is the only place this is used, so let's keep the logic all in one place.

File: packages/list-store/src/useSearchParamsUpdate.tsx [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  const [isPending, startTransition] = useTransition();
+  const state = useListStore();
+
+  useEffect(() => {
+    if (state.newDataNeeded && !isPending) {
+      startTransition(() => {
+        useListStore.setState({newDataNeeded: false});
```
[the remark this answers is not in the document]
Comment (barbarah) [reply packages/list-store/src/useSearchParamsUpdate.tsx:22]: `state` does not have a `setState` function. I could add a "startTransition" function to the store. Then I can do state.startTransition() and update the `newDataNeeded` in the store function.

File: packages/list-store/src/useSearchParamsUpdate.tsx
[the remark this answers is not in the document]
Comment (barbarah) [reply packages/list-store/src/useSearchParamsUpdate.tsx:1]: This is not needed, there is no jsx in this file. I will rename it.

File: packages/list-store/src/useListStore.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  selectedFilters: {[filterKey: string]: string[] | undefined};
+  setSelectedFilters: (key: string, value: string[]) => void;
+  setSortBy: (sortBy: SortBy) => void;
+  setQuery: (query: string) => void;
+  setPage: (direction: 1 | -1) => void;
+  composed: {
+    urlWithSearchParams: string;
+  };
+  newServerData: ({
```
[the remark this answers is not in the document]
Comment (barbarah) [reply packages/list-store/src/useListStore.ts:18]: I think it's important context that this setter is not a client action. All other setters are user actions that will result in collecting new data. Both types of actions have different side effects. A user action will also set `offset` and `newDataNeeded`. The `newServerData` will set `isInitialized`.

Somehow I want to make this context clear. Maybe the user actions also need to be renamed. I need some time to think about this.

File: packages/list-store/src/useSearchParamsUpdate.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (barbarah) [reply packages/list-store/src/useSearchParamsUpdate.tsx]: @sdevalk Thinking about the next names. Any thoughts?

User actions:

```
  filterClick
  sortClick / sortChange
  queryChange / queryTyping
  pageChange
```

API change

```
setNewData
```

File: packages/list-store/src/useListStore.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (barbarah) [reply packages/list-store/src/useListStore.ts:18]: @sdevalk Thinking about the next names. Any thoughts?

User actions:

```
  filterClick
  sortClick / sortChange
  queryChange / queryTyping
  pageChange
```

API change

```
setNewData
```

[the remark this answers is not in the document]
Comment (barbarah) [reply packages/list-store/src/useListStore.ts]: Decided on `[property]Change` for user actions. Because it sets more than just the property, so `set[Property]` can be confusing. And `setNewData` for new API data.

File: packages/list-store/src/useListStore.ts
[the remark this answers is not in the document]
Comment (barbarah) [reply packages/list-store/src/useListStore.ts:1]: I have added unit tests, only for the functions that do more than just setting data.

File: apps/researcher/cypress/e2e/home.cy.ts [hunk 7]
```diff
@@ -1,7 +1,88 @@
...
+    });
+
+    cy.getBySel('ownersFilter').within(() => {
+      cy.get('[type="checkbox"]').first().uncheck();
+    });
+
+    cy.getBySel('selectedFilter').should('have.length', 0);
+  });
+
+  it('removes a owner filter by deselecting it in the selected filter bar', () => {
```
Comment (sdevalk) [inline apps/researcher/cypress/e2e/home.cy.ts:42]: Minor: `removes an owner`

File: apps/researcher/src/lib/search-params.ts [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+});
+
+// Always return the sort values so the client knows how to sort.
+interface SearchOptionsWithRequiredSort extends SearchOptions {
+  sortBy: NonNullable<SearchOptions['sortBy']>;
+  sortOrder: NonNullable<SearchOptions['sortOrder']>;
+}
+
+// This function translates the search params to valid search options.
+// Next.js already separate the search query string into separates properties with string values.
```
Comment (sdevalk) [inline apps/researcher/src/lib/search-params.ts:76]: `// Next.js already separates the search query string into separate properties with string values.`?

File: packages/list-store/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,6 @@
+export * from './useListStore';
+export * from './sort';
+// Explicitly reexported client components to prevent error "Error: Unsupported Server Component type: undefined".
```
Comment (sdevalk) [inline packages/list-store/src/index.ts:3]: Minor: `Explicitly reexport client...`?

Would `export * from './client-list-store';` also work (to be consistent with the other exports)? Or do you explicitly do not want to do this?

File: packages/list-store/src/useListStore.test.ts [hunk 10]
```diff
@@ -0,0 +1,142 @@
+/**
+ * @jest-environment jsdom
+ */
+
+import {renderHook, act} from '@testing-library/react';
+import {useListStore} from './useListStore';
+import {SortBy} from './sort';
+
+describe('pageChange', () => {
+  it('will add the `limit` to the `offset` if the new `offset` is not greater than the `totalCount`', () => {
```
Comment (sdevalk) [inline packages/list-store/src/useListStore.test.ts:10]: For consistency with the other tests, use an active voice? For example: `it('adds the limit to...')`

File: packages/list-store/src/useListStore.test.ts [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    });
+
+    expect(result.current).toEqual(
+      expect.objectContaining({
+        ...newData,
+        isInitialized: true,
+      })
+    );
+  });
+  it('will not set values the user can edit if initialized', () => {
```
Comment (sdevalk) [inline packages/list-store/src/useListStore.test.ts:85]: Add a newline before the `it()`?

File: packages/list-store/src/useListStore.test.ts [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+      },
+    };
+
+    act(() => {
+      result.current.setNewData(newData);
+    });
+
+    expect(result.current).toEqual(expect.objectContaining(initializedData));
+  });
+  it('will set the `totalCount` if initialized', () => {
```
Comment (sdevalk) [inline packages/list-store/src/useListStore.test.ts:116]: Add a newline before the `it()`?

File: packages/list-store/src/useListStore.test.ts [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+    act(() => {
+      result.current.pageChange(-1);
+    });
+    expect(result.current.offset).toBe(0);
+  });
+});
+
+describe('setNewData', () => {
+  it('will set all data if `isInitialized is `false`', () => {
+    useListStore.setState({
```
[the remark this answers is not in the document]
Comment (barbarah) [reply packages/list-store/src/useListStore.test.ts:53]: I only set a variable if I need it later on. I don't need this variable later in this test. But if you prefer consistency, I will change it.

**Review by sdevalk — COMMENTED** [review]
Nice! Just some small comments :)

File: apps/researcher/cypress/e2e/home.cy.ts [hunk 14]
```diff
@@ -1,7 +1,88 @@
-describe('Research app homepage', () => {
-  // Replace with real tests
-  it('loads the page', () => {
+describe('Researcher homepage', () => {
+  it('loads an object list', () => {
```
[the remark this answers is not in the document]
Comment (barbarah) [reply apps/researcher/cypress/e2e/home.cy.ts:2]: I will change it to "shows an object list"

File: apps/researcher/src/lib/search-params.ts [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+  owners?: string;
+  types?: string;
+  subjects?: string;
+  query?: string;
+  offset?: string;
+  sortBy?: SortBy;
+}
+
+// Based on https://github.com/colinhacks/zod/issues/316#issuecomment-1024793482
+export function fallback<T>(value: T) {
```
[the remark this answers is not in the document]
Comment (barbarah) [reply apps/researcher/src/lib/search-params.ts:35]: Good question. No, I will remove the export.

**Review by sdevalk — APPROVED** [review]
👍🏻
